hamp_radar/serde_flight.py

```diff
@@ -38,66 +39,125 @@ def ppar_for_dataset(
         )
 
-    ppar = DatasetBlockGeometry(filename, mmbg)
+    return DatasetBlockGeometry(filename, mmbg)
+
+
+def generate_dbgs(pfgs: Iterable[PdsFileGeometry]) -> Iterable[DatasetBlockGeometry]:
+    """
+    Generates DatasetBlockGeometry intsances from a series of PdsFileGeometry instances
+
+    Args:
+        pfgs (Iterable[PdsFileGeometry]): An iterable of PdsFileGeometry instances.
+
+    Yields:
+        DatasetBlockGeometry: A DatasetBlockGeometry object for each mainblock in
+                              the PdsFileGeometry iterable.
+    """
+    for pfg in pfgs:
+        for mmbg in pfg.mainblocks:
+            if mmbg.subblocks[0].tag == "PPAR":
+                yield pparblock_for_dataset(pfg.filename, mmbg)
+            else:
+                yield DatasetBlockGeometry(pfg.filename, mmbg)
+
+
+def next_pparblock(blocks: Iterable[DatasetBlockGeometry]) -> DatasetBlockGeometry:
+    """
+    Finds and returns the next PPAR block from an (peekable) iterable of dataset blocks.
+
+    Function finds the "last of the first" PPAR block, i.e. it finds the first
+    occurance a PPAR in the list of DatasetBlockGeometry instances and if more
+    PPAR blocks follow consecutively after this one then the latest PPAR block
+    in the consecutive PPARs is returned.
+
+    Args:
+        blocks (Iterable[DatasetBlockGeometry]): Peekable iterable of dataset blocks to search through.
+
+    Returns:
+        DatasetBlockGeometry: The next PPAR block found in the iterable.
+
+    Raises:
+        IndexError: If no PPAR block is found by iterating over the dataset blocks.
+    """
+    import warnings
+
+    try:
+        while blocks.peek().mainblock.subblocks[0].tag != "PPAR":
+            # skip data if no PPAR block has yet been found
+            warnings.warn(
+                "Warning: data blocks before PPAR skipped",
+                UserWarning,
+            )
+            skipblock = next(blocks)
+    except IndexError as e:
+        e.add_note(f"No PPAR found in list of dataset blocks for {skipblock.filename}")
+        raise
+
+    while blocks.peek().mainblock.subblocks[0].tag == "PPAR":
+        ppar = next(blocks)
+        if not blocks:
+            warnings.warn(
+                "Warning: no data blocks appear after PPAR",
+                UserWarning,
+            )
+            break
+
     return ppar
 
 
+def next_datablocks(
+    blocks: Iterable[DatasetBlockGeometry],
+) -> List[DatasetBlockGeometry]:
+    """
+    Returns a list of dataset blocks from an (peekable) iterable for all the blocks
+    until a PPAR block is found of the iterable is exhausted.
+
+    Args:
+        blocks (Iterable[DatasetBlockGeometry]): An iterable of dataset blocks to search through.
+
+    Returns:
+        List[DatasetBlockGeometry]: The dataset blocks up to the next PPAR / end of the iterable.
+    """
+    data = []
+    while blocks and blocks.peek().mainblock.subblocks[0].tag != "PPAR":
+        data.append(next(blocks))
```

Comment:
Probably this should be a generator as well.

Reply:
it's a tiny unit I doubt we'll reuse so I included it within the function, I guess it could even be a lambda but maybe it's nice to be a bit more explicit